_posts/2019-09-30-introducing-arrow-flight.md

```diff
@@ -0,0 +1,260 @@
+Google has done on the problem).
+
+The best-supported way to use gRPC is to define services using an extended
+version of [Protocol Buffers][3] aka "Protobuf". A Protobuf plugin for gRPC
+generates a gRPC service that you can implement in your applications. RPC
+commands and data messages are serialized using the [Protobuf wire
+format][4]. Because we use "vanilla gRPC and Protocol Buffers", gRPC clients
+that are ignorant of the Arrow columnar format can still interact with Flight
+services and handle the Arrow data opaquely.
+
+The main data-related Protobuf type in Flight is called `FlightData`. Reading
+and writing Protobuf messages in general is not free, so we implemented some
+low-level optimizations in gRPC in both C++ and Java to do the following:
+
+* Generate the Protobuf wire format for `FlightData` including the Arrow record
+  batch being sent without going through any intermediate memory copying or
+  serialization steps.
+* Reconstruct a Arrow record batch from the Protobuf representation of
+  `FlightData` without any memory copying or deserialization.
+
+In a sense we are "having our cake and eat it, too". Flight implementations
+having these optimizations will have better performance, while naive gRPC
+clients talking to the Flight service and use a Protobuf library to deserialize
+`FlightData` (though with some performance penalty).
+
+As far as absolute speed, in our C++ data throughput benchmarks, we are seeing
+end-to-end TCP throughput in excess of 2-3GB/s on localhost without TLS
+enabled.
+
+```shell
+$ ./arrow-flight-benchmark --records_per_stream 100000000
+Bytes read: 12800000000
+Nanos: 3900466413
+Speed: 3129.63 MB/s
+
+$ ./arrow-flight-benchmark --records_per_stream 100000000
+Bytes read: 12800000000
+Nanos: 3631432266
+Speed: 3361.49 MB/s
+
+$ ./arrow-flight-benchmark --records_per_stream 100000000
+Bytes read: 12800000000
+Nanos: 4730784322
+Speed: 2580.34 MB/s
+```
+
+From this we can conclude that the machinery of Flight and gRPC adds relatively
+little overhead, and it suggests that many real-world applications of Flight
+will be bottlenecked on network bandwidth.
+
+## Horizontal Scalability: Parallel and Partitioned Data Access
+
+Many distributed database-type systems make use of a architectural pattern
+where the results of client requests are routed through a "coordinator" and
+sent to the client. Aside from the obvious efficiency issues of transporting a
+dataset multiple times on its way to a client, it also presents a scalability
+problem for getting access to very large datasets.
+
+We wanted Flight to enable systems to create horizontally scalable data
+services without this issue. A client request to a dataset using the
+`GetFlightInfo` RPC returns a list of **endpoints**, each of which contains a
+server location and a **ticket** to send that server in a `DoGet` request to
+obtain a part of the full dataset. To get access to the entire dataset, all of
+the endpoints must be consumed.
```

Comment:
Ah... I thought the locations were alternative locations. The batches in the dataset are not ordered?

Reply:
Correct, they are not alternatives. 

By default the streams are not ordered. You can use the `app_metadata` field in `FlightData` to include ordinal information if it's needed. I added a sentence to clarify